### PR TITLE
More trait changes

### PR DESCRIPTION
**Pull request by str4d**
This ports the remaining useful parts of the [`pasta_curves::arithmetic::FieldExt` trait](https://docs.rs/pasta_curves/0.4.1/pasta_curves/arithmetic/trait.FieldExt.html) onto `PrimeField` and two new traits (that are added because we can't use associated constants in trait method signatures yet).

### CODE REVIEW

File: ff_derive/src/lib.rs [hunk 1]
```diff
@@ -577,10 +590,11 @@ fn prime_field_constants_and_sqrt(
         };
 
     // Compute R^2 mod m
-    let r2 = biguint_to_u64_vec((&r * &r) % modulus, limbs);
+    let r2 = biguint_to_u64_vec(to_mont(r.clone()), limbs);
```
Comment (daira) [inline ff_derive/src/lib.rs:593]: Hmm. Yes I guess this is the same computation, but it's a bit odd. I think it might be more obvious to revert this change.

File: src/lib.rs [hunk 2]
```diff
@@ -191,7 +191,7 @@ pub trait Field:
     }
 }
 
-/// This represents an element of a prime field.
+/// This represents an element of a non-binary prime field.
```
Comment (daira) [inline src/lib.rs:194]: Non-binary representation ftw :-)

File: src/lib.rs [hunk 3]
```diff
@@ -241,6 +241,26 @@ pub trait PrimeField: Field + From<u64> {
...
+    ///
+    /// If you want to deterministically sample a field element representing a value, use
+    /// [`FromUniformBytes`] instead.
+    fn from_u128(v: u128) -> Self {
+        let lower = v as u64;
+        let upper = (v >> 64) as u64;
+        let mut tmp = Self::from(upper);
+        for _ in 0..64 {
+            tmp = tmp.double();
+        }
```
Comment (daira) [inline src/lib.rs:260]: Ideally I would like Field to support shifts (`x << w` $= x \cdot 2^w$ and `x >> w` $= x \cdot 2^{-w}$).

File: src/lib.rs [hunk 4]
```diff
@@ -310,6 +339,110 @@ pub trait PrimeField: Field + From<u64> {
...
+    /// It can be calculated by exponentiating [`Self::MULTIPLICATIVE_GENERATOR`] by `2^s`,
+    /// where `s` is [`Self::S`].
+    const DELTA: Self;
+}
+
+/// The subset of prime-order fields such that `(modulus - 1)` is divisible by `N`.
+///
+/// In these fields, there will be two valid choices of [`Self::ZETA`]. Similarly to
+/// [`PrimeField::MULTIPLICATIVE_GENERATOR`], the specific choice does not matter, as long
+/// as the choice is consistent across all uses of the field.
```
Comment (daira) [inline src/lib.rs:357]: I think that "two valid choices" only applies when $N = 3$. In other cases, there will be $N-1$ choices for prime $N$.

Our curve generation scripts have always chosen $\zeta$ for $N = 3$ to be the smaller of the two possibilities. I recommend this for interoperability.

```suggestion
/// If `N` is prime, there will be `N - 1` valid choices of [`Self::ZETA`]. Similarly to
/// [`PrimeField::MULTIPLICATIVE_GENERATOR`], the specific choice does not matter, as long
/// as the choice is consistent across all uses of the field. Choosing the value that is
/// smallest, when considered as an integer, may help to ensure consistency.
```

File: src/lib.rs [hunk 5]
```diff
@@ -310,6 +339,110 @@ pub trait PrimeField: Field + From<u64> {
...
+pub trait WithSmallOrderMulGroup<const N: u8>: PrimeField {
+    /// A field element of small multiplicative order $N$.
+    ///
+    /// The presense of this element allows you to perform (certain types of)
+    /// endomorphisms on some elliptic curves.
+    ///
+    /// It can be calculated using [SageMath] as
+    /// `GF(modulus).primitive_element() ^ ((modulus - 1) // N)`.
```
Comment (daira) [inline src/lib.rs:365]:
```suggestion
    /// `GF(modulus).primitive_element() ^ ((modulus - 1) // N)`.
    /// Choosing the value of order $N$ that is smallest, when considered
    /// as an integer, may help to ensure consistency.
```

File: src/lib.rs [hunk 6]
```diff
@@ -310,6 +339,110 @@ pub trait PrimeField: Field + From<u64> {
...
+/// let hash = blake2b(b"Some message");
+/// let val = Scalar::from_uniform_bytes(hash.as_array());
+/// # }
+/// ```
+///
+/// # Implementing `FromUniformBytes`
+///
+/// [`Self::from_uniform_bytes`] should always be implemented by interpreting the provided
+/// byte array as the little endian encoding of an integer, and then reducing that integer
+/// modulo the field modulus.
```
Comment (daira) [inline src/lib.rs:424]:
```suggestion
/// byte array as the little endian encoding of an unsigned integer, and then reducing that
/// integer modulo the field modulus.
```

File: src/lib.rs [hunk 7]
```diff
@@ -310,6 +339,110 @@ pub trait PrimeField: Field + From<u64> {
...
+/// type signature, and we do not want `PrimeField` to require a generic const parameter).
+/// However, this has the side-effect that `FromUniformBytes` can be implemented multiple
+/// times for different values of `N`. Most implementations of [`PrimeField`] should only
+/// need to implement `FromUniformBytes` trait for one value of `N` (chosen following the
+/// above considerations); if you find yourself needing to implement it multiple times,
+/// please [let us know about your use case](https://github.com/zkcrypto/ff/issues/new) so
+/// we can take it into consideration for future evolutions of the `ff` traits.
+pub trait FromUniformBytes<const N: usize>: PrimeField {
+    /// Returns a field element that is congruent to the provided little endian byte
+    /// representation of an integer.
```
Comment (daira) [inline src/lib.rs:444]:
```suggestion
    /// Returns a field element that is congruent to the provided little endian unsigned
    /// byte representation of an integer.
```

**Review by daira — APPROVED** [review]
utACK with suggestions.

File: src/lib.rs (hunk 4 above)
Comment (str4d) [reply src/lib.rs:357]: Good catch. The "two valid choices" language was from then this trait only represented N = 3 that FieldExt did, before I suggested we generalize it.

File: src/lib.rs (hunk 3 above)
Comment (str4d) [reply src/lib.rs:260]: This was issue #60, and no one disagreed with my counter-argument, so I closed the issue. If you have a motivating use case or argument, please share it there.

File: ff_derive/src/lib.rs (hunk 1 above)
Comment (str4d) [reply ff_derive/src/lib.rs:593]: Ah yes, this was caught in an automated refactor. I agree this specific case should be reverted.

Comment (str4d) [inline ff_derive/src/lib.rs:593]:
```suggestion
    let r2 = biguint_to_u64_vec((&r * &r) % modulus, limbs);
```

File: src/lib.rs (hunk 3 above)
Comment (daira) [reply src/lib.rs:260]: I hadn't noticed that had been closed; I'll reply there.

**Review by ebfull — APPROVED** [review]
